sphinx/builders/linkcheck.py

```diff
@@ -437,9 +449,30 @@ def _check_uri(self, uri: str, hyperlink: Hyperlink) -> tuple[str, str, int]:
             except HTTPError as err:
                 error_message = str(err)
 
-                # Unauthorised: the reference probably exists
+                # Unauthorized: the client did not provide required credentials
                 if status_code == 401:
-                    return 'working', 'unauthorized', 0
+                    if self._allow_unauthorized:
+                        deprecation_msg = (
+                            "\n---\n"
+                            "The linkcheck builder encountered an HTTP 401 "
+                            "(unauthorized) response, and will report it as "
+                            "'working' in this version of Sphinx to maintain "
+                            "backwards-compatibility."
+                            "\n"
+                            "This logic will change in Sphinx 8.0 which will "
+                            "report the hyperlink as 'broken'."
+                            "\n"
+                            "To explicitly continue treating unauthorized "
+                            "hyperlink responses as 'working', set the "
+                            "'linkcheck_allow_unauthorized' config option to "
+                            "``True``."
+                            "\n"
+                            "See sphinx-doc/sphinx#11433 for details."
+                            "\n---"
+                        )
+                        warnings.warn(deprecation_msg, RemovedInSphinx80Warning, stacklevel=1)
```

Comment:
Since the warning is quite long, should we emit it only the first time? (I don't know whether those warnings are only emitted once or not by default or if we are filtering them at the very top level).

Reply:
By default, warnings are de-duplicated based on message, source module and source line.

So for the following code:

```py
import warnings

def warning(msg):
    warnings.warn(msg)

warning('hello warning 1')
warning('hello warning 1')
warning('hello warning 1')
warning('hello warning 2')
warning('hello warning 3')
warning('hello warning 3')
```

...we'll see the following output:

```
/home/jka/foo.py:4: UserWarning: hello warning 1
  warnings.warn(msg)
/home/jka/foo.py:4: UserWarning: hello warning 2
  warnings.warn(msg)
/home/jka/foo.py:4: UserWarning: hello warning 3
  warnings.warn(msg)
```

(six method calls being de-duplicated to three distinct warning outputs)

However if we run the code with `PYTHONWARNINGS=always`, it'll instead display all of the six emitted warnings - no de-duplication.

(all of the above is generic Python behaviour - none of this is specific to, or overridden by, Sphinx)

Reply:
Oh then it's fine!

Reply:
Ok, thanks - I did double-check this again a few moments ago to make sure; and yep, when using a `sphinx-build -b linkcheck doc <output-dir>`, a `warnings.warn` message (with a static/constant `message` value) was only emitted once despite the code being reached multiple times.

Reply:
I think it's because the default warning's filter is set to "once" for Python (or something like that).

Reply:
I suppose it's _possible_ that someone has a large number of HTTP 401s _and_ wants to consider them as `working` _and_ wants to set `PYTHONWARNINGS=always` -- and for them I suppose it's true that this will produce a large amount of log output.

Hopefully it's reasonable to allow them to file a bugreport / contact us if so?  It would be possible to adjust fairly easily but I feel like it might be over-adjusting for a case that I don't know exists.

Reply:
> I suppose it's _possible_ that someone has a large number of HTTP 401s _and_ wants to consider them as `working` _and_ wants to set `PYTHONWARNINGS=always` -- and for them I suppose it's true that this will produce a large amount of log output.

Self-correction: that middle condition is the default with this change (HTTP 401 continues to mean `working` by default), so that should be omitted because it would be true for most deployments.

Reply:
Yeah. Let's stick to the simple case. If there set "always", then they more or less know what they are doing also.